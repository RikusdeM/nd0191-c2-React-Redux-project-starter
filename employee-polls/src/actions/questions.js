import {
  saveQuestion,
  saveQuestionAnswer,
  saveQuestionUserUpdate,
} from "../utils/api";
import {
  updateUserAddAnswer,
  updateUserRemoveAnswer,
  updateUserQuestions,
} from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const UNANSWER_QUESTION = "UNANSWER_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddQuestion(questionsText) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    // dispatch(showLoading());
    //Todo: Re-add loading

    const updateBackend = async () => {
      try {
        const question = await saveQuestion({
          optionOneText: questionsText.optionOneText,
          optionTwoText: questionsText.optionTwoText,
          author: authedUser,
        });
        const updateUserQ = await saveQuestionUserUpdate({
          authedUser: authedUser,
          qid: question.id,
        });

        dispatch(addQuestion({ [question.id]: question }));
        dispatch(
          updateUserQuestions({
            authedUser: authedUser,
            id: question.id,
          })
        );
      } catch (error) {
        console.error("An error occurred: ", error);
        alert("Could not add Question, please try again")
      }
    };

    return updateBackend();
  };
}

export function answerQuestion({ authedUser, id, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    id,
    answer,
  };
}

export function unanswerQuestion({ authedUser, id, answer }) {
  return {
    type: UNANSWER_QUESTION,
    authedUser,
    id,
    answer,
  };
}

export function handleAnswerQuestion({ authedUser, id, answer }) {
  return (dispatch) => {
    // const { authedUser } = getState();
    dispatch(answerQuestion({ authedUser, id, answer }));
    dispatch(updateUserAddAnswer({ authedUser, id, answer }));

    return saveQuestionAnswer({
      authedUser: authedUser,
      qid: id,
      answer: answer,
    }).catch((e) => {
      console.warn("Error in handleAnswerQuestion: ", e);
      dispatch(unanswerQuestion({ authedUser, id, answer }));
      dispatch(updateUserRemoveAnswer({ authedUser, id }));
      alert("Error answering Poll, please try again");
    });
  };
}
