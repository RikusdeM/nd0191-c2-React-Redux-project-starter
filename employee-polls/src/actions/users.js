export const RECEIVE_USERS = "RECEIVE_USERS";
export const UPDATE_USER_ADD_ANSWER = "UPDATE_USER_ADD_ANSWER";
export const UPDATE_USER_REMOVE_ANSWER = "UPDATE_USER_REMOVE_ANSWER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function updateUserAddAnswer({ authedUser, id, answer }) {
  return {
    type: UPDATE_USER_ADD_ANSWER,
    authedUser,
    id,
    answer,
  };
}
export function updateUserRemoveAnswer({ authedUser, id, answer }) {
  return {
    type: UPDATE_USER_REMOVE_ANSWER,
    authedUser,
    id,
  };
}
