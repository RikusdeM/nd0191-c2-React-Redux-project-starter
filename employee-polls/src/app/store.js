import { configureStore, Tuple } from "@reduxjs/toolkit";
import reducer from "../reducers";
import { thunk } from "redux-thunk";
import logger from "../middleware/logger";

export const store = configureStore({
    reducer: reducer,
    middleware: () => new Tuple(thunk, logger),
  });