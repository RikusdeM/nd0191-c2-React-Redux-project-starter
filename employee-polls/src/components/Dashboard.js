import { connect } from "react-redux";
import QuestionList from "./QuestionList";
import "bootstrap/dist/css/bootstrap.css";
import Dropdown from "react-bootstrap/Dropdown";
import DropdownButton from "react-bootstrap/DropdownButton";
import { useState } from "react";

const Dashboard = ({
  answeredQuestionsIds,
  unansweredQuestionsIds,
  authedUser,
}) => {
  const QUESTIONS = {
    UNANSWERED: 0,
    ANSWERED: 1,
    ALL: 2,
  };

  const [dropDown, setDropDown] = useState(QUESTIONS.UNANSWERED);  

  return (
    <div>
      <DropdownButton id="dropdown-basic-button" title="Show Polls">
        <Dropdown.Item          
          disabled={dropDown === QUESTIONS.UNANSWERED}
          onClick={(e) => setDropDown(QUESTIONS.UNANSWERED)}
        >
          New Questions
        </Dropdown.Item>
        <Dropdown.Item disabled={dropDown === QUESTIONS.ANSWERED} onClick={(e) => setDropDown(QUESTIONS.ANSWERED)}>
          Done
        </Dropdown.Item>
        <Dropdown.Item disabled={dropDown === QUESTIONS.ALL} onClick={(e) => setDropDown(QUESTIONS.ALL)}>
          All
        </Dropdown.Item>
      </DropdownButton>

      {dropDown === QUESTIONS.UNANSWERED ? (
        <div className="container content-center">
          <QuestionList
            name={"New Questions"}
            questionsIds={unansweredQuestionsIds}
          />
        </div>
      ) : null}

      {dropDown === QUESTIONS.ANSWERED ? (
        <div className="container content-center">
          <QuestionList name={"Done"} questionsIds={answeredQuestionsIds} />
        </div>
      ) : null}

      {dropDown === QUESTIONS.ALL ? (
        <div className="container content-center">
          <div className="row dashboard-padding">
            <div className="col">
              <QuestionList
                name={"New Questions"}
                questionsIds={unansweredQuestionsIds}
              />
            </div>
          </div>
          <div className="row">
            <div className="col">
              <QuestionList name={"Done"} questionsIds={answeredQuestionsIds} />
            </div>
          </div>
        </div>
      ) : null}
    </div>
  );
};

const mapStateToProps = ({ questions, authedUser, users }) => {
  const sortQuestions = (keys) => {
    return keys.sort((a, b) => questions[b].timestamp - questions[a].timestamp);
  };

  const answeredQuestionsIds = Object.keys(users[authedUser].answers);

  const unansweredQuestionsIds = Object.keys(questions).filter(
    (qId) => !answeredQuestionsIds.includes(qId)
  );

  return {
    answeredQuestionsIds: sortQuestions(answeredQuestionsIds),
    unansweredQuestionsIds: sortQuestions(unansweredQuestionsIds),
    authedUser,
  };
};

export default connect(mapStateToProps)(Dashboard);
