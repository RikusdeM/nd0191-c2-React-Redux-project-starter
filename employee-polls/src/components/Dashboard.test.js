import { render } from "@testing-library/react";
import Dashboard from "./Dashboard";
import { expect, describe, it } from "@jest/globals";
import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { store } from "../app/store";
import { handleInitialData } from "../actions/shared";
import { setAuthedUser } from "../actions/authedUser";

describe("Dashboard", () => {
  it("will match the snapshot", async () => {
    /* Configure the store with the same static initial Data and authed user. 
    The component should render the same when the initial data with 
    the same authedUser is used.
    */
    const authedUser = "sarahedo";

    await store.dispatch(handleInitialData());
    await store.dispatch(setAuthedUser(authedUser));

    const dashboard = render(
      <MemoryRouter initialEntries={[`/`]}>
        <Provider store={store}>
          <Dashboard />
        </Provider>
      </MemoryRouter>
    );
    expect(dashboard).toMatchSnapshot();
  });
});
