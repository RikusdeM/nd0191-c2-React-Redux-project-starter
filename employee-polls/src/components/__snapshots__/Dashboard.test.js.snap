// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard will match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="dropdown"
        >
          <button
            aria-expanded="false"
            class="dropdown-toggle btn btn-primary"
            id="dropdown-basic-button"
            type="button"
          >
            Show Polls
          </button>
        </div>
        <div
          class="container content-center"
        >
          <div>
            <h2>
              New Questions
            </h2>
            <div
              class="container"
            >
              <div
                class="Container container"
              >
                <div
                  class="Row row"
                >
                  <div
                    class="Col col-md-3"
                  >
                    <div>
                      <div
                        class="card"
                        style="width: 15rem;"
                      >
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            mtsamis
                          </h5>
                          <h6
                            class="card-subtitle mb-2 text-muted"
                          >
                            9:16:PM | 4/30/2017
                          </h6>
                          <button
                            class="btn btn-sm btn-block btn-primary"
                            type="button"
                          >
                            Show
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="Col col-md-3"
                  >
                    <div>
                      <div
                        class="card"
                        style="width: 15rem;"
                      >
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            tylermcginnis
                          </h5>
                          <h6
                            class="card-subtitle mb-2 text-muted"
                          >
                            1:09:PM | 3/15/2017
                          </h6>
                          <button
                            class="btn btn-sm btn-block btn-primary"
                            type="button"
                          >
                            Show
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="dropdown"
      >
        <button
          aria-expanded="false"
          class="dropdown-toggle btn btn-primary"
          id="dropdown-basic-button"
          type="button"
        >
          Show Polls
        </button>
      </div>
      <div
        class="container content-center"
      >
        <div>
          <h2>
            New Questions
          </h2>
          <div
            class="container"
          >
            <div
              class="Container container"
            >
              <div
                class="Row row"
              >
                <div
                  class="Col col-md-3"
                >
                  <div>
                    <div
                      class="card"
                      style="width: 15rem;"
                    >
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          mtsamis
                        </h5>
                        <h6
                          class="card-subtitle mb-2 text-muted"
                        >
                          9:16:PM | 4/30/2017
                        </h6>
                        <button
                          class="btn btn-sm btn-block btn-primary"
                          type="button"
                        >
                          Show
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="Col col-md-3"
                >
                  <div>
                    <div
                      class="card"
                      style="width: 15rem;"
                    >
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          tylermcginnis
                        </h5>
                        <h6
                          class="card-subtitle mb-2 text-muted"
                        >
                          1:09:PM | 3/15/2017
                        </h6>
                        <button
                          class="btn btn-sm btn-block btn-primary"
                          type="button"
                        >
                          Show
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
